File: Retainic/Language.swift
//
//  Language.swift
//  Retainic
//
//  A curated set of languages the user can choose from.
//

import Foundation

struct Language: Identifiable, Hashable {
    let code: String
    let name: String

    var id: String { code }

    static let all: [Language] = [
        Language(code: "en", name: "English"),
        Language(code: "es", name: "Spanish"),
        Language(code: "zh", name: "Chinese"),
        Language(code: "ja", name: "Japanese"),
        Language(code: "ko", name: "Korean"),
    ]

    static func named(_ code: String) -> Language? {
        all.first { $0.code == code }
    }

    var displayName: String { name }

    /// The language's name written in the given UI language code, e.g. "ja"
    /// renders English as "英語". Falls back to the English `name`.
    func displayName(in uiCode: String) -> String {
        let locale = Locale(identifier: Language.localeIdentifier(for: uiCode))
        guard let localized = locale.localizedString(forLanguageCode: code) else { return name }
        return localized.capitalized(with: locale)
    }

    /// The language's name written in itself (its autonym), e.g. "Español",
    /// "简体中文", "日本語". Falls back to the English `name`.
    var autonym: String {
        let identifier = Language.localeIdentifier(for: code)
        let locale = Locale(identifier: identifier)
        guard let localized = locale.localizedString(forLanguageCode: identifier) else { return name }
        return localized.capitalized(with: locale)
    }

    /// Maps an app language code to a full locale identifier for bundle lookup.
    static func localeIdentifier(for code: String) -> String {
        code == "zh" ? "zh-Hans" : code
    }

    /// The `Locale` for an app language code, used to drive the interface.
    static func locale(for code: String) -> Locale {
        Locale(identifier: localeIdentifier(for: code))
    }

    /// The best supported language for the current device, defaulting to English.
    static var systemDefault: String {
        let preferred = Locale.preferredLanguages.first ?? "en"
        let code = Locale(identifier: preferred).language.languageCode?.identifier ?? "en"
        return all.contains { $0.code == code } ? code : "en"
    }
}

File: android/app/src/main/java/com/retainic/app/i18n/Language.kt
package com.retainic.app.i18n

import java.util.Locale

/**
 * The curated set of interface / study languages. Ported from Language.swift.
 * Display names come from the JVM's locale data, matching the iOS behaviour of
 * rendering a language's name in the chosen UI language (or in itself).
 */
data class Language(val code: String, val name: String) {
    /** The language's name written in the given UI language code. */
    fun displayName(uiCode: String): String {
        val target = Locale.forLanguageTag(localeTag(code))
        val ui = Locale.forLanguageTag(localeTag(uiCode))
        val localized = target.getDisplayLanguage(ui)
        if (localized.isBlank()) return name
        return localized.replaceFirstChar { it.titlecase(ui) }
    }

    /** The language's name written in itself (its autonym), e.g. "Español". */
    val autonym: String get() = displayName(code)

    companion object {
        val all: List<Language> = listOf(
            Language("en", "English"),
            Language("es", "Spanish"),
            Language("zh", "Chinese"),
            Language("ja", "Japanese"),
            Language("ko", "Korean"),
        )

        fun named(code: String): Language? = all.firstOrNull { it.code == code }

        /** Maps an app language code to a BCP-47 tag for locale lookup. */
        fun localeTag(code: String): String = if (code == "zh") "zh-Hans" else code

        /** The best supported language for the device, defaulting to English. */
        fun systemDefault(): String {
            val code = Locale.getDefault().language
            return if (all.any { it.code == code }) code else "en"
        }
    }
}
